### PR TITLE
OPSD: on-policy self-distillation trainer

### PR DESCRIPTION
**Pull request by kashif**
Implements On-Policy Self-Distillation from https://huggingface.co/papers/2601.18734 (official code: https://github.com/siyan-zhao/OPSD) as a third self-distillation trainer in `trl/experimental/opsd`, same self-contained design and conventions as sdft and sdpo.

One model, two roles: the student samples an on-policy completion from the bare problem, the teacher (same network, frozen at the initial weights by default) scores it conditioned on the problem plus the ground truth `solution`. The loss is the full vocabulary JSD along the student trajectory, with the paper's pointwise per vocabulary entry KL clipping (`distillation_kl_clip`) so style tokens don't dominate the signal.

All the sibling trainer features carry over: base/live/ema teacher, vLLM colocate and server, teacher scoring on the server, liger, peft, and separate `teacher_chat_template_kwargs` for the non thinking student / thinking teacher pairing the paper recommends.

17 unit tests plus smoke runs on serial, zero3 (vLLM colocate and server with `use_teacher_server`), fsdp2 with PEFT, all training cleanly.

SDFT and SDPO now use the shared callbacks module too: their `teacher_sync.py` copies are removed, so they pick up the sharding-aware EMA callback (gathered reads under ZeRO-3, DTensor handling under FSDP2) and the same ZeRO-3 + PEFT guard. Smoke tested fsdp2 EMA + LoRA for both trainers.

---

> [!NOTE]
> **Medium Risk**
> Introduces a large new training loop (generation buffering, dual forwards, vLLM teacher server) and changes shared EMA callback behavior for SDFT/SDPO under distributed training, though patterns mirror existing self-distillation trainers.
> 
> **Overview**
> Adds **On-Policy Self-Distillation (OPSD)** as `trl.experimental.opsd` (`OPSDConfig`, `OPSDTrainer`, loss helpers), aligned with SDFT/SDPO: on-policy student rollouts, teacher conditioned on `privileged_context` (ground-truth solution), configurable divergence modes (`full_logits` default with **per-vocab KL clip**), base/live/EMA teachers, vLLM generation, optional **teacher logprobs via vLLM server**, Liger fused JSD, and PEFT. Ships `examples/opsd_math`, docs (`opsd_trainer.md`, paper index, toctree), and a broad unit test suite.
> 
> **SDFT** and **SDPO** are updated in parallel: teacher EMA/sync callbacks move to **`trl.experimental.callbacks`** (fix import path; **ZeRO-3 gather** and **FSDP2 DTensor** handling when syncing LoRA teacher adapters), teacher prompts use **untruncated** tokenization so problem text is not dropped, plus shared guards (PEFT + ZeRO-3, `parallelism_config` CP/SP unsupported), QLoRA bf16 adapter cast, Liger/`logit_scale` checks, and small config typing fixes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 99f4214a7b56fa8677d1ec59d421c8351f9b9b5a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/opsd/opsd_trainer.py [hunk 1]
```diff
@@ -0,0 +1,1282 @@
...
+    def __init__(
+        self,
+        model: str | PreTrainedModel | nn.Module,
+        args: OPSDConfig | None = None,
+        train_dataset: Dataset | IterableDataset | None = None,
+        eval_dataset: Dataset | IterableDataset | dict[str, Dataset | IterableDataset] | None = None,
+        processing_class: PreTrainedTokenizerBase | ProcessorMixin | None = None,
+        callbacks: list[TrainerCallback] | None = None,
+        optimizers: tuple[torch.optim.Optimizer | None, torch.optim.lr_scheduler.LambdaLR | None] = (None, None),
+        peft_config: PeftConfig | None = None,
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/opsd/opsd_trainer.py]: fixed, quoted it the way the main trainers do. fwiw the merged sdft and sdpo trainers have the same unquoted annotation, so importing those without peft hits the same NameError, worth a follow up there.

**Comment by LeonEricsson** [comment]
OPSD appears to be **very** similar to SDFT: sample student rollouts from the prompt, score those same completions under a teacher-conditioned prompt using privileged_context, and minimize a token-level divergence between student and teacher distributions. On a first pass the OPSD-specific differences I found was the default objective (full_logits forward KL), distillation_kl_clip, teacher-top-k support when using top-k mode, and the more specific teacher prompt/ground-truth-solution framing.

should we consider implementing OPSD as a specific configuration or specialization of SDFT rather than as a fully separate trainer?

File: trl/experimental/callbacks.py
Comment (LeonEricsson) [inline trl/experimental/callbacks.py:1]: i'm happy/fine with this move into a shared module, but I'd like @qgallouedec to weigh in as well.

File: trl/experimental/opsd/opsd_trainer.py [hunk 2]
```diff
@@ -0,0 +1,1296 @@
...
+        self.trainer = trainer
+
+    def _stringify_privileged_context(self, privileged_context: Any) -> str:
+        if privileged_context is None:
+            raise ValueError(
+                "`privileged_context` must not be None for self-distillation teacher prompt construction."
+            )
+        if isinstance(privileged_context, str):
+            return privileged_context
+        return str(privileged_context)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/opsd/opsd_trainer.py:165]: good catch, fixed in 833815f0. opsd now pulls the text out of a conversational privileged_context the same way sdft does, instead of calling str() on the message list.

File: trl/experimental/opsd/opsd_trainer.py [hunk 3]
```diff
@@ -0,0 +1,1296 @@
...
+            model_init_kwargs["device_map"] = None
+        self.teacher_model = create_model_from_path(get_config_model_id(self.model.config), **model_init_kwargs)
+        self.teacher_model.requires_grad_(False)
+        self.teacher_model.eval()
+        if self.is_deepspeed_enabled:
+            self.teacher_model = prepare_deepspeed(self.teacher_model, self.accelerator)
+        elif self.is_fsdp_enabled:
+            self.teacher_model = prepare_fsdp(self.teacher_model, self.accelerator)
+        else:
+            self.teacher_model = self.accelerator.prepare_model(self.teacher_model, evaluation_mode=True)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/opsd/opsd_trainer.py:624]: this is intentional and matches sdft/sdpo (and the grpo/dpo reference model). we build the teacher from the path instead of deepcopying the student because under zero3 the student params are already sharded, so a deepcopy would clone empty shards. the base/ema teacher assumes a student loaded from that path, same as the rest of the family.

File: trl/experimental/opsd/opsd_trainer.py [hunk 4]
```diff
@@ -0,0 +1,1307 @@
...
+                tokenize=True,
+                return_dict=True,
+                **(self.chat_template_kwargs if chat_template_kwargs is None else chat_template_kwargs),
+            )
+            prompt_ids = tokenized["input_ids"]
+        else:
+            prompt_ids = self.processing_class(text=prompts)["input_ids"]
+        if self.max_prompt_length is not None:
+            prompt_ids = [ids[-self.max_prompt_length :] for ids in prompt_ids]
+        return prompt_ids
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/opsd/opsd_trainer.py:843]: good catch, fixed in 526b39be. the teacher prompt now goes through a _tokenize_prompts_untruncated path so the leading problem is never dropped, while the student generation prompt keeps the left truncation. this matches how sdpo already does it. applied the same fix to sdft since it had the identical issue.

**Comment by behroozazarkhalili** [comment]
A data point for the "specialization of SDFT vs separate trainer" question @LeonEricsson raised, from diffing the two branches at the loss level.

**On the loss.** This PR adds a new `trl/experimental/opsd/loss_utils.py` that is a near-duplicate of `main`'s `trl/experimental/sdft/loss_utils.py`. SDFT already has `compute_divergence` (the same alpha-divergence: forward KL, reverse KL, and the logsumexp-mixture skew for intermediate alpha), `compute_full_logit_self_distillation_loss`, `compute_topk_self_distillation_loss`, `add_tail_bucket`, plus the config surface `distillation_mode in {sampled_token, full_logits, topk_logits}`, `distillation_alpha`, `distillation_add_tail`, `distillation_topk`, and `distillation_is_clip` (the clipped importance-sampling correction). Diffing OPSD's copy against SDFT's, the algorithmic deltas I could find are just two:

1. `compute_topk_self_distillation_loss` takes the top-k support from the **teacher** (`torch.topk(teacher_logits)`) rather than SDFT's **student** support (`torch.topk(student_logits)`), a one-line difference in which tensor is passed to `torch.topk`.
2. An optional per-vocabulary-entry `kl_clip` (`kl.clamp(max=kl_clip)`) added to OPSD's `compute_divergence`; SDFT's `compute_divergence` on `main` has no such parameter.

Both read like they would fit as SDFT knobs (e.g. a `topk_support="teacher"` option and a `distillation_kl_clip` arg) on the local-teacher path, rather than a forked loss file.

**In fairness to the other side.** OPSD does carry non-loss machinery SDFT does not: the `SolutionTeacherContextBuilder` / ground-truth-solution teacher-prompt framing, the EMA-teacher option, and the untruncated teacher-prompt tokenization. But the trainer method structure (`_compute_self_distillation_loss`, `_compute_server_distillation_loss`, `_compute_teacher_student_logits`, `_compute_liger_loss`, `_get_teacher_context_for_self_distillation`) mirrors SDFT's almost one-to-one, so those pieces look SDFT-shaped too. And OPSD adds a `paper_index.md` entry, so if it maps to a distinct published method there is a discoverability argument for a standalone trainer.

I do not think this is mine to decide. It is a genuine tradeoff between the repo's stated preference for self-contained, duplicated-on-purpose trainers and the ~2600 lines of near-duplicate this adds. Just wanted to make the loss-level overlap precise so the config-vs-trainer call can be made on the actual deltas rather than a first-pass impression. @qgallouedec, since you own that duplication-vs-specialization line for the trainers, this one probably wants your read.

File: docs/source/_toctree.yml [hunk 5]
```diff
@@ -133,6 +133,8 @@
     title: PPO
   - local: prm_trainer
     title: PRM
+  - local: opsd_trainer
+    title: OPSD
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply docs/source/_toctree.yml:137]: I checked this one against the file before it gets acted on, and I think it is inverted.

Counting every entry in `docs/source/_toctree.yml` at this commit:

```
indent histogram for '- local:'  : {2: 62, 4: 3}
indent histogram for its 'title:': {4: 62, 6: 3}
opsd_trainer (local_indent, title_indent) = (2, 4)
```

So 62 of the 65 entries use 2 spaces for `- local:` and 4 for `title:`, and only 3 use the 4/6 pattern the comment describes as the house style. `opsd_trainer` is at 2/4, which is the dominant convention and identical to its immediate neighbours:

```yaml
  - local: prm_trainer
    title: PRM
  - local: opsd_trainer
    title: OPSD
  - local: sdft_trainer
    title: SDFT
```

Re-indenting OPSD to 4/6 would move it out of this list and change the document structure, so this is worth leaving as is. Flagging it here rather than in the PR thread so nobody spends time on it.

**Comment by qgallouedec** [comment]
Thanks @kashif, this is clearly a lot of careful work, sorry it sat that long

The problem is scope. At 3.4k lines across 21 files there are really three PRs here:

1. the OPSD trainer itself
2. moving SDFT and SDPO onto the shared callbacks and deleting `sdpo/teacher_sync.py`
3. the vLLM changes, batched `update_named_params` plus 66 lines in `scripts/vllm_serve.py`

2 and 3 are the parts that can affect people already using SDFT and SDPO, and right now they are buried inside a new trainer review. Can you pull them out?
They should land quickly on their own, and OPSD then gets reviewed as a self-contained addition.

One heads up on 3!: #6765 replaces our vLLM server with vLLM's own, so anything added to `scripts/vllm_serve.py` will not survive it. Worth rebasing on that first.

It is also conflicting with main at the moment.

File: trl/experimental/callbacks.py [hunk 6]
```diff
@@ -101,46 +101,59 @@ def __init__(
...
-        set_peft_model_state_dict(model, teacher_state, adapter_name=self.teacher_adapter_name)
-
-        model.set_adapter(adapter_name)
+        student_state = self._get_student_state_dict()
+        self.shadow_weights = {k: torch.zeros_like(v) for k, v in student_state.items()}
 
-        self.shadow_weights = {k: v.clone().zero_() for k, v in teacher_state.items()}
+        if self.teacher_adapter_name not in model.peft_config:
+            model.add_adapter(self.teacher_adapter_name, self.teacher_adapter_config)
+        self._write_teacher_weights(model)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/callbacks.py:156]: checked this empirically: `add_adapter("teacher", ...)` leaves the new adapter at `requires_grad=False` (PEFT only marks the *active* adapter trainable), and `_write_teacher_weights`/`_initialize_teacher_adapter` never call `set_adapter`, so it's never activated either. Ran the exact `PEFTAdapterEMACallback` flow end-to-end: 0 teacher-adapter params trainable, 8/8 student-adapter params trainable, active adapter stays "default" throughout. The teacher forward during training goes through `use_adapter(...)`, which switches to "teacher" only for that forward and restores the student's adapter (and its requires_grad state) before backward/optimizer.step(). So this doesn't reproduce — leaving it open in case I'm missing a code path, happy to take a repro.

File: trl/experimental/sdft/sdft_config.py [hunk 7]
```diff
@@ -196,10 +200,10 @@ class SDFTConfig(_BaseConfig):
         },
     )
     disable_dropout: bool = field(
-        default=True,
+        default=False,
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/sdft/sdft_config.py]: good catch — reverted `disable_dropout` back to `True` for opsd/sdft/sdpo in d9421488's follow-up. Copied distillation's default flip without checking: distillation's teacher is always a separate, persistently-`.eval()`'d model, so dropout on the student never touches it. Here, `teacher_model_kind` in {`live`, `base`+peft, `ema`+pure-lora} reuses `self.model` as the teacher with no separate eval-mode toggle, so `disable_dropout=False` would make the teacher forward stochastic during training exactly as you described. Left a comment on the field explaining why it has to stay `True` here, unlike GRPO/DistillationTrainer.

File: trl/experimental/opsd/loss_utils.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            torch.stack([student_log_probs + torch.log(1 - alpha_t), teacher_log_probs + torch.log(alpha_t)]),
+            dim=0,
+        )
+        kl_teacher = F.kl_div(mixture, teacher_log_probs, reduction="none", log_target=True)
+        kl_student = F.kl_div(mixture, student_log_probs, reduction="none", log_target=True)
+        kl = torch.lerp(kl_student, kl_teacher, alpha)
+    if kl_clip is not None:
+        # Pointwise per-vocabulary-entry clip (OPSD): caps high-divergence style tokens before the vocabulary sum.
+        kl = kl.clamp(max=kl_clip)
+    return kl.sum(-1)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/opsd/loss_utils.py]: fixed — moved the vocabulary sum before the clamp (`kl = kl.sum(-1); if kl_clip is not None: kl = kl.clamp(max=kl_clip)`), so the clip now caps the already-non-negative per-position KL instead of individual per-vocab-entry terms. Same fix as ca0e6e1c on huggingface/trl#6237, which had the identical bug. Applied to opsd/sdft/sdpo (only opsd exposes kl_clip today, but the pattern would've bitten sdft/sdpo too if they ever add it).

File: trl/experimental/opsd/opsd_trainer.py [hunk 9]
```diff
@@ -0,0 +1,1387 @@
...
+                scaled = getattr(config, "logit_scale", 1.0) not in (None, 1.0)
+                softcapped = getattr(config, "final_logit_softcapping", None) is not None
+                if scaled or softcapped:
+                    raise ValueError(
+                        f"`use_liger_kernel=True` is incompatible with the {name} model's `logit_scale` / "
+                        f"`final_logit_softcapping` (e.g. Cohere / Gemma models): the Liger fused JSD loss reads "
+                        f"`lm_head.weight` directly and cannot apply them, so it would optimize a different "
+                        f"objective than the model's real forward. Set `use_liger_kernel=False` to use the chunked "
+                        f"loss, which applies both."
+                    )
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/opsd/opsd_trainer.py]: good catch — the config was being read after `_setup_teacher_model()` already wrapped the teacher (DeepSpeed/FSDP), so `.config` could be the wrong object for the separate-teacher-copy path. Fixed by checking compatibility inline inside `_setup_teacher_model`, right where the teacher is still unwrapped, before each of the deepspeed/fsdp/plain-prepare branches. Also restructured to match the mainline DistillationTrainer's actual implementation of this check (`model.config.get_text_config()` for VLM support, `output_multiplier` fallback for Muse Glimmer) instead of my own ad-hoc version — verified with a standalone smoke test. Applied to opsd/sdft/sdpo.

**Comment by qgallouedec** [comment]
@kashif the aug 28 commits look right but the split from aug 20 hasn't happened, it's still 3.3k lines across 21 files and it grew.

Can you pull the liger `logit_scale` fix out first and open it on its own? It fixes sdft and sdpo, which are on main today, and reading `teacher_model.config` after the deepspeed wrap means the check has never fired for anyone using deepspeed. That shouldn't wait on this PR. I'll merge it as soon as it's up.

Also @LeonEricsson asked whether opsd is sdft with different defaults, that's still unanswered and it's upstream of the split: if most of opsd is sdft with a different objective default, PR 1 is a config change rather than a new trainer. But maybe I'm missing something, can you answer that first, then split?

Thanks!

**Comment by qgallouedec** [comment]
Closing to keep the review queue tidy, it's been sitting for a while. Reopen whenever, the work itself is good.

Two things worth carrying over. Scope is still 3.3k lines across 18 files covering four independent changes: the trainer, the shared-callbacks move, the liger `logit_scale` fix and `models/utils.py`. And @LeonEricsson's june question shapes the split: `opsd/loss_utils.py` is 31 lines from `sdft/loss_utils.py`, `opsd_config.py` adds two fields and drops two, `opsd_trainer.py` is 251 changed lines out of 1400. So OPSD lands as SDFT with two knobs and different defaults, making the first piece a config change rather than a new trainer.

The liger `logit_scale` fix doesn't need to wait on any of that: it fixes sdft and sdpo on main, and reading `teacher_model.config` after the deepspeed wrap means the check has never fired on deepspeed.
